compiler/codegen/CCData.hpp

```diff
@@ -0,0 +1,207 @@
+       *
+       * Unique Object Representations
+       *
+       * Structs and classes usually have padding bytes which are going to be included in their memory representations.
+       * Even if the padding is at the end of the instance, sizeof() will include it. Calculating a key based on the memory
+       * representation is going to include the padding, which is incorrect. If the key includes the padding, two otherwise
+       * equal objects will result in different keys if their padding bytes differ.
+       *
+       * This also applies to floats and doubles because of various "don't care" (DC) bits in the binary representation.
+       * Logically equal float values might have different bit patterns and would result in different keys.
+       *
+       * Types that don't have any padding or DC bits have unique object representations in C++ standard parlance.
+       *
+       * @tparam T The type of the value to create the key from. The type must have a unique object representation. A static assertion checks for the std::has_unique_object_representations<T> type trait on capable compilers.
+       * @param[In] value The value to create the key from.
+       * @return The key.
+       */
+      template <typename T>
+      static key_t key(const T value);
+
+      /**
+       * @brief Creates a key_t from the given buffer of data. The entire buffer will be used as input, including any unused bits/bytes that you may not be aware of. See key(const T value) for info on why this is important to note.
+       *
+       * @param[In] data A pointer to the buffer of data.
+       * @param[In] sizeBytes The size (in bytes) of the data.
+       * @return The key.
+       */
+      static key_t key(const uint8_t * const data, const size_t sizeBytes);
+
+      /**
+       * @brief Creates a key_t from the given null-terminated C string.
+
+       * @param[In] data A pointer to the string.
+       * @return The key.
+       * \TODO: Decide if we need this; useful for string literals, but susceptible to bugs/attack because of the unbounded string.
+       */
+      static key_t key(const char * const str);
+
+      /**
+       * @brief Constructs a CCData.
+
+       * @param[In] sizeBytes The amount of data (in bytes) that CCData can hold.
+       */
+      CCData(const size_t sizeBytes);
+
+      /**
+       * @brief Constructs a CCData and accepts a pointer to a memory buffer that will be used to hold the data.
+       *
+       * The lifetime of the memory buffer must exceed the lifetime of the resulting CCData object. The memory buffer will still be valid after the CCData object is destructed (i.e. the memory buffer will not be freed, do it yourself).
+       *
+       * @param[In] storage A pointer to a memory buffer that CCData will use.
+       * @param[In] sizeBytes The amount of data (in bytes) that the buffer contains.
+       */
+      CCData(uint8_t * const storage, const size_t sizeBytes);
```

Comment:
The second one is the one that's actually used in the compiler. The first was convenient for testing, I could just create `CCData` instances without repeating myself, but it's actually not necessary and eliminating it would get rid of the constructor and anti-pattern you mentioned elsewhere. I'll remove it.